### PR TITLE
[STM32] Feature/can3 peripheral

### PR DESCRIPTION
**Pull request by FalcioneE**
Some STM32 microcontrollers include 3 CAN peripherals. Previously, Libcanard only supported the use of CAN1 and CAN2. This pull request would add support for CAN3 to Libcanard.

Changes:
_internal_bxcan.h: Include memory location of CAN3 config register
canard_stm32.h: Add build config macro for CAN3
canard_stm32.c: Make filter config generic so it will work for either CAN1 or CAN3

From Pavel: Some parameters have to be configured against the CAN1 instance, so replacing CANARD_STM32_CAN1 with BXCAN will break the driver (at least for some MCUs). This is covered in the Reference Manual.
Can you direct me to the reference manual please? 

This driver code works on my STM32F765VIT.

### CODE REVIEW

**Review by pavel-kirienko — CHANGES_REQUESTED** [review]
> Can you direct me to the reference manual please?

This is from the RM for STM32F10x:

image

The RM for STM32F7 says the same thing:

image

Did you actually test that the filters are functional? The RM says it shouldn't work. Even if it does, the driver would be reliant on undocumented behavior, which is usually unacceptable.

The patch also needs some formatting cleanup.

**Comment by FalcioneE** [comment]
This is the ref manual I've been working from:
https://www.st.com/resource/en/reference_manual/dm00224583-stm32f76xxx-and-stm32f77xxx-advanced-arm-based-32-bit-mcus-stmicroelectronics.pdf

which includes this snippet: 
image


CAN1 and CAN2 are multiplexed, and the configuration of some CAN2 settings are dependant on the CAN1 setup. CAN3 is entirely separate. I believe I've left the UAVCAN driver code such that the necessary CAN1 parameters are set up when CAN2 is selected.

**Comment by pavel-kirienko** [comment]
Indeed, this special case requires special treatment. Observe though that you are accessing `F*` registers (such as `FM1R` etc) via the `BXCAN` macro, which resolves to CAN2 when CAN2 is selected, which breaks the driver. You could introduce an additional reference to the master CAN instance to solve this, such that it would resolve to CAN1 when either CAN1 or CAN2 are used; otherwise, it would resolve to CAN3.

Please also fix the formatting issues.

**Comment by FalcioneE** [comment]
New commit pushed, let me know your thoughts. 

I think I caught most of the formatting differences, let me know if there are other things you need me to fix.

File: stm32/libcanard/canard_stm32.h [hunk 1]
```diff
@@ -19,13 +19,29 @@ extern "C"
...
 #if !defined(CANARD_STM32_USE_CAN2)
 # define CANARD_STM32_USE_CAN2                                  0
 #endif
 
+/**
+ * Set this build config macro to 1 to use CAN3 instead of CAN1
+ * Setting this parameter when CAN3 is not available may not be detected at compile time!
+ */
+#if !defined(CANARD_STM32_USE_CAN3)
+# define CANARD_STM32_USE_CAN3                                  1
```
Comment (pavel-kirienko) [inline stm32/libcanard/canard_stm32.h:42]:
```suggestion
# define CANARD_STM32_USE_CAN3                                  0
```

I don't imagine this configuration to be very common, so let the user choose the correct one explicitly.

File: stm32/libcanard/canard_stm32.h [hunk 2]
```diff
@@ -19,13 +19,29 @@ extern "C"
 #endif
 
 /**
- * Set this build config macro to 1 to use CAN2 instead of CAN1, if available.
+ * Set this build config macro to 1 to use CAN1
+ * Setting this parameter when CAN1 is not available may not be detected at compile time!
```
Comment (pavel-kirienko) [inline stm32/libcanard/canard_stm32.h:23]:
```suggestion
```

AFAIK there are no chips where CAN1 is not available.

File: stm32/libcanard/canard_stm32.c [hunk 3]
```diff
@@ -11,10 +11,15 @@
...
+# define BXCAN                                                  CANARD_STM32_CAN3
+# define FILTER_CONFIG_BXCAN                                    CANARD_STM32_CAN3
+#elif CANARD_STM32_USE_CAN2
 # define BXCAN                                                  CANARD_STM32_CAN2
-#else
+# define FILTER_CONFIG_BXCAN                                    CANARD_STM32_CAN1
+#elif CANARD_STM32_USE_CAN1
 # define BXCAN                                                  CANARD_STM32_CAN1
+# define FILTER_CONFIG_BXCAN                                    CANARD_STM32_CAN1
 #endif
```
Comment (pavel-kirienko) [inline stm32/libcanard/canard_stm32.c:23]: I suggest we add an `ifndef` check below that makes sure that BXCAN is defined. If not, the user forgot to define any of `CANARD_STM32_USE_CAN*`.

File: stm32/libcanard/canard_stm32.c [hunk 4]
```diff
@@ -460,7 +465,7 @@ int16_t canardSTM32ConfigureAcceptanceFilters(const CanardSTM32AcceptanceFilterC
      * First we disable all filters. This may cause momentary RX frame losses, but the application
      * should be able to tolerate that.
      */
-    CANARD_STM32_CAN1->FA1R = 0;
+    BXCAN->FA1R = 0;
```
Comment (pavel-kirienko) [inline stm32/libcanard/canard_stm32.c:468]:
```suggestion
    FILTER_CONFIG_BXCAN->FA1R = 0;
```

File: stm32/libcanard/canard_stm32.c [hunk 5]
```diff
@@ -552,10 +557,10 @@ int16_t canardSTM32ConfigureAcceptanceFilters(const CanardSTM32AcceptanceFilterC
 #else
             i;
 #endif
-        CANARD_STM32_CAN1->FilterRegister[filter_index].FR1 = id;
-        CANARD_STM32_CAN1->FilterRegister[filter_index].FR2 = mask;
+        BXCAN->FilterRegister[filter_index].FR1 = id;
+        BXCAN->FilterRegister[filter_index].FR2 = mask;
```
Comment (pavel-kirienko) [inline stm32/libcanard/canard_stm32.c:561]:
```suggestion
        FILTER_CONFIG_BXCAN->FilterRegister[filter_index].FR1 = id;
        FILTER_CONFIG_BXCAN->FilterRegister[filter_index].FR2 = mask;
```

File: stm32/libcanard/canard_stm32.c [hunk 6]
```diff
@@ -552,10 +557,10 @@ int16_t canardSTM32ConfigureAcceptanceFilters(const CanardSTM32AcceptanceFilterC
...
 
-        CANARD_STM32_CAN1->FA1R |= 1U << filter_index;      // Enable
+        BXCAN->FA1R |= 1U << filter_index;      // Enable
```
Comment (pavel-kirienko) [inline stm32/libcanard/canard_stm32.c:563]:
```suggestion
        FILTER_CONFIG_BXCAN->FA1R |= 1U << filter_index;      // Enable
```

File: stm32/libcanard/canard_stm32.c [hunk 7]
```diff
@@ -246,33 +251,33 @@ int16_t canardSTM32Init(const CanardSTM32CANTimings* const timings,
...
     CANARD_STM32_CAN1->FilterRegister[CANARD_STM32_NUM_ACCEPTANCE_FILTERS].FR1 = 0;
     CANARD_STM32_CAN1->FilterRegister[CANARD_STM32_NUM_ACCEPTANCE_FILTERS].FR2 = 0;
     CANARD_STM32_CAN1->FA1R = (1 << CANARD_STM32_NUM_ACCEPTANCE_FILTERS);  // One filter enabled
 #else
-    CANARD_STM32_CAN1->FilterRegister[0].FR1 = 0;
-    CANARD_STM32_CAN1->FilterRegister[0].FR2 = 0;
-    CANARD_STM32_CAN1->FA1R = 1;                                        // One filter enabled
+    BXCAN->FilterRegister[0].FR1 = 0;
+    BXCAN->FilterRegister[0].FR2 = 0;
+    BXCAN->FA1R = 1;                                        // One filter enabled
```
Comment (pavel-kirienko) [inline stm32/libcanard/canard_stm32.c:277]: I suggest we use `FILTER_CONFIG_BXCAN` here for clarity. The logic would not change but the intent should be more obvious.

**Review by pavel-kirienko — CHANGES_REQUESTED** [review]

**Comment by FalcioneE** [comment]
Added ifndef to check that some CAN peripheral has been selected.

File: stm32/libcanard/canard_stm32.c [hunk 8]
```diff
@@ -246,33 +255,33 @@ int16_t canardSTM32Init(const CanardSTM32CANTimings* const timings,
...
     CANARD_STM32_CAN1->FilterRegister[CANARD_STM32_NUM_ACCEPTANCE_FILTERS].FR1 = 0;
     CANARD_STM32_CAN1->FilterRegister[CANARD_STM32_NUM_ACCEPTANCE_FILTERS].FR2 = 0;
     CANARD_STM32_CAN1->FA1R = (1 << CANARD_STM32_NUM_ACCEPTANCE_FILTERS);  // One filter enabled
 #else
-    CANARD_STM32_CAN1->FilterRegister[0].FR1 = 0;
-    CANARD_STM32_CAN1->FilterRegister[0].FR2 = 0;
-    CANARD_STM32_CAN1->FA1R = 1;                                        // One filter enabled
+    FILTER_CONFIG_BXCAN->FilterRegister[0].FR1 = 0;
+    FILTER_CONFIG_BXCAN->FilterRegister[0].FR2 = 0;
+    BXCAN->FA1R = 1;                                        // One filter enabled
```
Comment (pavel-kirienko) [inline stm32/libcanard/canard_stm32.c:281]: Why not `FILTER_CONFIG_BXCAN`?

File: stm32/libcanard/canard_stm32.c [hunk 9]
```diff
@@ -11,10 +11,19 @@
...
 # define BXCAN                                                  CANARD_STM32_CAN2
-#else
+# define FILTER_CONFIG_BXCAN                                    CANARD_STM32_CAN1
+#elif CANARD_STM32_USE_CAN1
 # define BXCAN                                                  CANARD_STM32_CAN1
+# define FILTER_CONFIG_BXCAN                                    CANARD_STM32_CAN1
+#endif
+
+#ifndef BXCAN
+#error BXCAN not defined. Make a selection for BXCAN using the build config macros in canard_stm32.h.
```
Comment (pavel-kirienko) [inline stm32/libcanard/canard_stm32.c:26]:
```suggestion
# error BXCAN not defined. Make a selection for BXCAN using the build config macros in canard_stm32.h.
```

**Review by pavel-kirienko — CHANGES_REQUESTED** [review]
Thanks. You can request a review explicitly when ready by clicking this in the upper-right part of the page:

image

**Review by pavel-kirienko — APPROVED** [review]

**Pull request merged**